contracts/starknet/lib/session_key.cairo

```diff
@@ -0,0 +1,159 @@
+        alloc_locals
+        let (eth_address) = SessionKey_owner_store.read(session_public_key)
+        with_attr error_message("Session does not exist"):
+            assert_not_zero(eth_address)
+        end
+        StarkEIP191.verify_session_key_revoke_sig(r, s, salt, session_public_key)
+        revoke(session_public_key)
+        return ()
+    end
+
+    func revoke_with_owner_sig{
+        syscall_ptr : felt*,
+        range_check_ptr,
+        pedersen_ptr : HashBuiltin*,
+        bitwise_ptr : BitwiseBuiltin*,
+        ecdsa_ptr : SignatureBuiltin*,
+    }(r : Uint256, s : Uint256, v : felt, salt : Uint256, session_public_key : felt):
+        alloc_locals
+        let (eth_address) = SessionKey_owner_store.read(session_public_key)
+        with_attr error_message("Session does not exist"):
+            assert_not_zero(eth_address)
+        end
+        EIP712.verify_session_key_revoke_sig(r, s, v, salt, eth_address, session_public_key)
+        revoke(session_public_key)
+        return ()
+    end
+
+    func revoke_with_owner_tx{syscall_ptr : felt*, range_check_ptr, pedersen_ptr : HashBuiltin*}(
+        session_public_key : felt
+    ):
+        alloc_locals
+        let (eth_address) = SessionKey_owner_store.read(session_public_key)
+        with_attr error_message("Session does not exist"):
+            assert_not_zero(eth_address)
+        end
+        let (commit_array : felt*) = alloc()
+        assert commit_array[0] = eth_address
+        assert commit_array[1] = session_public_key
+        let (commit_hash) = HashArray.hash_array(2, commit_array)
+
+        # Checks that hash maches a commit and that the commit was created by the correct address
+        EthTx.consume_commit(commit_hash, eth_address)
+
+        revoke(session_public_key)
+
+        return ()
+    end
+
+    # Returns owner of a session key if it exists, otherwise throws
+    func get_owner{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        session_public_key : felt
+    ) -> (eth_address : felt):
+        let (eth_address) = SessionKey_owner_store.read(session_public_key)
+        with_attr error_message("Session does not exist"):
+            assert_not_zero(eth_address)
+        end
+
+        let (end_timestamp) = SessionKey_end_timestamp_store.read(session_public_key)
+        let (current_timestamp) = get_block_timestamp()
+        with_attr error_message("Session has ended"):
+            assert_le(current_timestamp, end_timestamp)
+        end
+        return (eth_address)
+    end
+end
+
+func register{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    eth_address : felt, session_public_key : felt, session_duration : felt
+):
+    SessionKey_owner_store.write(session_public_key, eth_address)
+    let (current_timestamp) = get_block_timestamp()
+    SessionKey_end_timestamp_store.write(session_public_key, current_timestamp + session_duration)
```

Comment:
Should we check for overflow here? What happens if `current_timestamp + session_duration` overflows? the number will be either negative, or `< current_timestamp`? This means it simply won't work, so might as well throw an error here right?

Reply:
yeah nice spot. could also enforce a max session duration?

Reply:
I just put an overflow check for now, some people might just want an indefinite session key to be active so seems unnecessary to limit it.

``    with_attr error_message("Overflow in Session duration, use smaller value"):
        assert_le(current_timestamp, end_timestamp)
    end
```